test/Phlexy/Lexer/Stateless/WithCapturingGroupsTest.php

```diff
@@ -7,11 +7,35 @@
 class WithCapturingGroupsTest extends \Phlexy\Lexer\TestAbstract {
     public function createLexerFactory() {
         return new \Phlexy\LexerFactory\Stateless\WithCapturingGroups(
-            new \Phlexy\LexerDataGenerator
+            new \Phlexy\LexerDataGenerator,
+            false
         );
     }
 
     public function provideTestLexing() {
         return $this->getTestsWithCapturingGroups();
     }
-}
\ No newline at end of file
+
+    public function testNamedGroups() {
+        $lexerFactory = new \Phlexy\LexerFactory\Stateless\WithCapturingGroups(
+            new \Phlexy\LexerDataGenerator,
+            true
+        );
+
+        $lexer = $lexerFactory->createLexer(
+            array(
+                '\s+'          => 0,
+                '\$(?<varName>\w+)'      => 1,
+                '(?<intPart>\d+)\.(?<fractionalPart>\d+)' => 2
```

Comment:
Does this also work if two tokens use the same named group? I think that might result in a PCRE pattern as the patterns are merged and we're (IIRC) not using capture-reset groups here.

Reply:
Yes, it's a problem.

Originally, lexer definition was like
```php
$lexer = $factory->createLexer(array(
    '[^{}\\\\]+'                        => Token::T_PLAIN_TEXT,
    '\{\\?(?::(?<index>\d+))\}' => Token::T_INDEXED_PLACEHOLDER,
    '\{(?<value>(?:\\\\.|[a-zA-Z0-9_.-]*)*)\}' => Token::T_NAMED_PLACEHOLDER,
    '\\\\(?<escaped>.)'                 => Token::T_ESCAPED_SEQUENCE,
));
```
In short, it's like PSR-3 format but with regular named placeholders (`{foo}`) and indexed (`{?:42}`) + escaping.

I tried to make some research:

1. `DEFINE` + `(?|...)`:

    ```php
    $pattern = '~
        \\G
        (?(DEFINE)
            (?<plaintext>[^{}\\\\]+)
            (?<indexed>\{\\?(?::(?<index>\d+))\})
            (?<named>\{(?<value>(?:\\\\.|[a-zA-Z0-9_.-]*)*)\})
            (?<escaped>\\\\(?<escaped_char>.))
        )
        (?|(?&plaintext)|(?&indexed)|(?&named)|(?&escaped))
        ~x';
    ```

    ```
    [8] =>
      array(1) {
        [0] =>
        string(5) "{foo}"
      }
      [9] =>
      array(1) {
        [0] =>
        string(5) "test "
      }
    ```

    It seems like there is no way to get subgroups from named groups in `DEFINE` block.

2. `(?|` without named groups:

    ```php
    $pattern = '~
        \\G
        (?|
            (?:[^{}\\\\]+)
            |
            (?:\{\\?(?::(\d+))\})
            |
            (?:\{((?:\\\\.|[a-zA-Z0-9_.-]*)*)\})
            |
            (?:\\\\(.))
        )
        ~x';
    ```

    ```
      [8] =>
      array(2) {
        [0] =>
        string(5) "{foo}"
        [1] =>
        string(3) "foo"
      }
      [9] =>
      array(1) {
        [0] =>
        string(5) "test "
      }
      [10] =>
      array(2) {
        [0] =>
        string(6) "{?:42}"
        [1] =>
        string(2) "42"
      }
    ```

    In this case I couldn't find way to provide token id: offset or named group. 

3. `(?|...)` + `J` modifier. I don't understand why it fails even if all groups are `(?:` and `(?<...>`: 
    ```
    $pattern = '~
    \\G
    (?|
        (?<plaintext>[^{}\\\\]+)
        |
        (?<indexed>\{\\?(?::(?:\d+))\})
        |
        (?<named>\{(?:(?:\\\\.|[a-zA-Z0-9_.-]*)*)\})
        |
        (?<escaped>\\\\(?:.))
    )
    ~xJ';
    ```

    → `Compilation failed: different names for subpatterns of the same number are not allowed at offset`.

But technically possible to rename groups, then rename them back at token object creation stage, but it's hard work.
Well, I'll probably close a bit later.

P.S. By the way, did you test performance of `preg_replace_callback()` vs `preg_match_all()`? They seem interchangeable here.

Reply:
I found out `(*MARK:foo)` syntax, so I tried experimental lexer using `preg_match_all`: https://github.com/nikic/Phlexy/compare/master...unkind:exprimental-preg-match-all (it also contains changes from this PR, I was too lazy).

It shows a slightly better results in benchmarks (against `UsingPregReplace`):
```diff
Timing lexing of CVS data:
Took 0.34519600868225 seconds (Phlexy\Lexer\Stateless\Simple)
Took 0.26229000091553 seconds (Phlexy\Lexer\Stateless\WithCapturingGroups)
Took 0.23753499984741 seconds (Phlexy\Lexer\Stateless\WithoutCapturingGroups)
Took 0.2250280380249 seconds (Phlexy\Lexer\Stateless\UsingPregReplace)
+Took 0.20805191993713 seconds (Phlexy\Lexer\Stateless\UsingPregMatchAll)

Timing alphabet lexing of all "a":
Took 0.3294689655304 seconds (Phlexy\Lexer\Stateless\Simple)
Took 0.35999298095703 seconds (Phlexy\Lexer\Stateless\WithCapturingGroups)
Took 0.32758617401123 seconds (Phlexy\Lexer\Stateless\WithoutCapturingGroups)
Took 0.3077449798584 seconds (Phlexy\Lexer\Stateless\UsingPregReplace)
+Took 0.28351402282715 seconds (Phlexy\Lexer\Stateless\UsingPregMatchAll)

Timing alphabet lexing of all "z":
Took 0.57571291923523 seconds (Phlexy\Lexer\Stateless\Simple)
Took 0.12331295013428 seconds (Phlexy\Lexer\Stateless\WithCapturingGroups)
Took 0.12075114250183 seconds (Phlexy\Lexer\Stateless\WithoutCapturingGroups)
Took 0.11755514144897 seconds (Phlexy\Lexer\Stateless\UsingPregReplace)
+Took 0.056530952453613 seconds (Phlexy\Lexer\Stateless\UsingPregMatchAll)

Timing alphabet lexing of random string:
Took 0.81303811073303 seconds (Phlexy\Lexer\Stateless\Simple)
Took 0.2578718662262 seconds (Phlexy\Lexer\Stateless\WithCapturingGroups)
Took 0.24983787536621 seconds (Phlexy\Lexer\Stateless\WithoutCapturingGroups)
Took 0.23963904380798 seconds (Phlexy\Lexer\Stateless\UsingPregReplace)
+Took 0.14482712745667 seconds (Phlexy\Lexer\Stateless\UsingPregMatchAll)
```
plus output is more readable from my point of view:
```
  69997 =>
  array (
    0 => 2,
    1 => 5000,
    2 => ',',
  ),
  69998 =>
  array (
    0 => 0,
    1 => 5000,
    2 => 'stuff',
  ),
```
vs
```
  69997 =>
  array (
    'full' => ',',
    'token' => 'token2',
    'line' => 5000,
    'matches' =>
    array (
    ),
  ),
  69998 =>
  array (
    'full' => 'stuff',
    'token' => 'token0',
    'line' => 5000,
    'matches' =>
    array (
    ),
  ),
```
Produced PCRE is 
```
~
\G
(?|
    (?:[^",\r\n]+)(*MARK:token0)
    |
    (?:"[^"\\]*(?:\\.[^"\\]*)*")(*MARK:token1)
    |
    (?:,)(*MARK:token2)|(?:\r?\n)(*MARK:token3)
    |
    [\S\s](*MARK:error)
)
~Jx
```
(added whitespaces for readability).

But I suspect performance is better not because of `preg_match_all`, but different pattern, need to check later.

Also, it explicitly forbids named groups in token regex'es:
```php
// Naive assertion
if (preg_match('/(?<!\\\\)(?<group>\(\?<[a-zA-Z_][a-zA-Z0-9_]*>)/', $regex, $match)) {
    throw new LogicException(sprintf('Named groups are forbidden in token regex; %s was found', $match['group']));
}

return sprintf('(?:%s)(*MARK:token%d)', $regex, $tokenId);
```
It would be cool to have PCRE Parser for PHP that produces AST, you know. Some of existing ones are not working.

But even without it, I suppose it's possible to replace `(?<...>` with `(` with remembering number of subgroup within token's regex and later replace all indexes with remembered group names.

I have no idea how it fits in your project, because I see too many changes that I want to do. I mean, it's kinda painful to add features in project for 5.3. :)

Reply:
`*MARK` should definitely be the best way to do something like this nowadays. In fact, I added this feature in PHP 5.6 (iirc) pretty much exactly for this use-case.

I think as the first step it would make sense to bump the minimum version requirement of this library to PHP 7.1. This library was written back when PHP 5.3 was still recent, but it certainly makes no sense to support this now.

With the minimum version raised, we will be able to use `*MARK`, which should make things both faster and easier to evolve.